include/mata/number-predicate.hh

```diff
@@ -0,0 +1,250 @@
+//
+// Created by Lukáš Holík on 06.12.2022.
+//
+
+#pragma once
+
+#ifndef LIBMATA_NUMBER_PREDICATE_HH
+#define LIBMATA_NUMBER_PREDICATE_HH
+
+#include <vector>
+#include <mata/ord-vector.hh>
+
+namespace Mata {
+    namespace Util {
+
+        template <class Number> class OrdVector;
+
+        template<typename Number>
+        class NumPredicate {
+        private:
+            mutable std::vector<bool> predicate = {};
+            mutable std::vector <Number> elements = {};
+            mutable bool elements_exact = true;
+            mutable bool elements_watched = true; //invariant: implies that elements overapproximate the true values
+            //TODO: if it is never used with elements_watched = false, then we can simplify.
+
+            using const_iterator = typename std::vector<Number>::const_iterator;
+
+            void prune_elements() const {
+                Number new_pos = 0;
+                for (Number orig_pos = 0; orig_pos < elements.size(); ++orig_pos) {
+                    if (predicate[elements[orig_pos]]) {
+                        elements[new_pos] = elements[orig_pos];
+                        ++new_pos;
+                    }
+                }
+                elements.resize(new_pos);
+                elements_exact = true;
+            }
+
+            void compute_elements() const {
+                elements.clear();
+                for (Number q = 0; q < predicate.size(); ++q) {
+                    if (predicate[q])
+                        elements.push_back(q);
+                }
+                elements_exact = true;
+            }
+
+            void update_elements() const {
```

Comment:
I would go with `update_satisfiable_elements` and explanation, that elements are satisfiable when they satisfy the predicate.

Reply:
Comment at the beginning could solve it.?

Reply:
I think, if we named all the functions and properties more carefully, it would need no comment. That is preferable in my opinion, though I think comments help as well.

Reply:
done